reVX/setbacks/setbacks.py

```diff
@@ -0,0 +1,148 @@
+        distinguished in the local regulations file. In particular,
+        this string (or list of strings) should match the values under
+        the ``Feature Type`` column of the setback regulations file
+        corresponding to this feature. For example, specifying
+        ``feature_type="property line"`` would mean that this setback
+        calculator looks for ``"property line"`` regulations (i.e.
+        parcel setbacks) in the input regulations file. This input is
+        case insensitive (i.e. "ROADS", "Roads", "roads", and "rOAds"
+        are all equivalent), and spaces, dashes, and underscores are
+        interchangeable (i.e. "property line", "property_line", and
+        "property-line" are all equivalent).
+    buffer_type : str, optional
+        Name of the buffer type to use for this setback calculation. See
+        :attr:`~reVX.setbacks.base.BUFFERS` for all available options -
+        this input must be a key from that dictionary. In most cases,
+        the default positive buffer is suitable.
+        By default, ``"default"``.
+    feature_filter_type : str, optional
+        Name of the feature filter type to use for this setback
+        calculation. Typically, features like structures or parcels are
+        filtered using "centroid", which only applies setback
+        regulations to features whose centroid is within the county with
+        the ordinance. On the other hand, features such as roads,
+        railroads transmission lines, etc., are filtered using "clip",
+        which clips the feature to the county boundaries before applying
+        a setback. This value must be a key in the
+        :attr:`~reVX.setbacks.base.FEATURE_FILTERS` dictionary.
+        By default, ``"centroid"``.
+    feature_subtypes_to_exclude : list | set, optional
+        A list of strings specifying the feature subtypes that should be
+        excluded from the local regulations calculation.
+        By default, ``None``.
+    num_features_per_worker : int, optional
+        Number of features each worker (CPU processor) on a node
+        processes. By default, ``10,000``.
+    """
+    if isinstance(feature_type, str):
+        feature_type = [feature_type]
+
+    feature_type = {_format_str(name) for name in feature_type}
+
+    if buffer_type not in BUFFERS:
+        msg = ("Unknown buffer type specified: {!r}. Must be one of {}"
+               .format(buffer_type, list(BUFFERS)))
+        logger.error(msg)
+        raise ValueError(msg)
+
+    if feature_filter_type not in FEATURE_FILTERS:
+        msg = ("Unknown feature filter type specified: {!r}. Must be one of {}"
+               .format(feature_filter_type, list(FEATURE_FILTERS)))
+        logger.error(msg)
+        raise ValueError(msg)
+
+    feature_subtypes_to_exclude = set(feature_subtypes_to_exclude or set())
+    num_features_per_worker = int(max(0, num_features_per_worker) or 10_000)
+
+    class Setbacks(AbstractBaseSetbacks):
```

Comment:
Interesting syntax putting the class inside the function. I feel like there has to be a cleaner way to do this... Maybe consider having AbstractBaseSetbacks be non-abstract and decorated as a [dataclass](https://docs.python.org/3/library/dataclasses.html)?

Also add a return docstring, especially with such a funky output haha

Reply:
Yea interesting thought. I'm not sure how simple it would be to decorate with a dataclass since `AbstractBaseSetbacks` inherits from another abstract class. However, making it non-abstract is a really interesting idea. This function call is essentially a class factory, but I could move the factory logic into a `Setbacks` class method. That would make the syntax something like 
```
return Setbacks.new(feature_type, feature_subtypes_to_exclude, buffer_type, feature_filter_type, num_features_per_worker)
```

I quite like that actually. Would that address the issue in your opinion?

Reply:
Well if the concrete setback class inherits the abstract class and just adds some class attributes, i would think we should make AbstractBaseSetbacks -> Setbacks where Setbacks is a data class (decorated with @dataclass or not). If all of the class attributes are required, just make them positional attrs. 

Class method that returns an instantiated Setbacks class would be a nice way to build and return the class. 

HMMM i guess is the main thing that I'm not understanding that you need the class returned and not an instance of the class?

Reply:
Right, this function is a class factory, so the return value is a class which can later be instantiated into an object. The reason this code is written this way is historical: In the past, each type of setback feature had it's own class implementation derived from a BaseSetback class. I wanted to put the power to generate new type of setback calculations into the users hands, so I consolidated the boilerplate of six child classes derived from BaseSetbacks into a single function that can generate the derived classes on demand.

I think it would technically be possible to move these attributes to the initializer. Again, the reason it wasn't done this way before is purely historical. Unfortunately this also means that making this change could be non trivial - the AbstractBaseSetbacks class would have to be renamed and refactored to allow extra initializer parameters, the parameter parsing logic would have to be moved somewhere within the class, and all code that instantiated/interacted with the derived setbacks classes (i.e. the majority of the setbacks tests) would have to be refactored to match the new init signature. This is all doable in theory, but I wonder if the effort is worthwhile in practice. If you think it is, please let me know and I can look into making this change.

In the meantime, I updated the syntax to be more functional in nature. Perhaps this alleviates some of the pain of the ugly old syntax?

Reply:
Sounds good, two things: camel case class name with strip, and ~reVX.setbacks.base.AbstractBaseSetbacks